### DISCUSSION TITLE
Assign mesh operation

### DISCUSSION THREAD

**Discussion by aknsj** [ansys/pyaedt#4886] [Q&A] [answered]
Hi all, I am using AEDT version 2024R1 and pyaedt version 0.9.2, which helps to automate the simulation process.
I have come across a small problem. The issue is that I want to select some solids and then assign mesh operation to them as shown below.
![image](https://github.com/ansys/pyaedt/assets/163245306/5dc1b457-2ef5-4661-a9ec-9964ffc6e3b1)

The parameters that need to be modified are shown below,, which are Level_min, Level_max and Local_Mesh_Parameter.
![image](https://github.com/ansys/pyaedt/assets/163245306/52634a63-7f5d-4ceb-a451-28bfac83fc54)
![image](https://github.com/ansys/pyaedt/assets/163245306/73edb090-7aa1-4c45-9c19-862f1c189f95)

I would be glad to know which function should be used to implement the assign mesh region function. Looking forward to your reply!

**Comment by lorenzovecchietti** [comment, marked as answer]
Hi, you have to modify manually the properties of a mesh operation

```
import pyaedt
from collections import OrderedDict

ipk=pyaedt.Icepak(specified_version="2024.1")
ipk.modeler.create_box([0,0,0], [1,1,1], name="Box1")
ipk.mesh.assign_mesh_level({"Box1":2})
ipk.mesh.meshoperations[0].props.update(OrderedDict({
 "MaxLevel": "3",
 "MinLevel": "1",
 "Local Mesh Parameters Enabled": True,
 "X count": "1.5",
 "Y count": "2.5",
 "Z count": "3"
}))
ipk.mesh.meshoperations[0].props.pop("Level")
ipk.mesh.meshoperations[0].update()
```

**Reply by aknsj** [reply]
Thanks for your answer. I tried the script you provided and successfully set mesh operation to one solid,which helped me. Further, I tried two methods to set mesh operation to multiple solids , but both methods raise the following error.
```
ipk.mesh.meshoperations[0].props.pop("Level")
KeyError: 'Level'
```

The first method is looping statement with the following code.
```
import pyaedt
from collections import OrderedDict
for solid_name in ipk.modeler.solid_names:
    if solid_name.lower().startswith("copper"):
        ipk.mesh.assign_mesh_level({solid_name:2})
        ipk.mesh.meshoperations[0].props.update(OrderedDict({
            "MaxLevel": "3",
            "MinLevel": "2",
            "Local Mesh Parameters Enabled": True,
            "X count": "2",
            "Y count": "2",
            "Z count": "2"
        }))
        ipk.mesh.meshoperations[0].props.pop("Level")
        ipk.mesh.meshoperations[0].update()
```
The second method is to construct a dictionary with the following code.
```
import pyaedt
from collections import OrderedDict
mesh_level_dict = {}
for solid_name in ipk.modeler.solid_names:
    if solid_name.lower().startswith("copper"):
        mesh_level_dict[solid_name] = 2
if mesh_level_dict!=0:
    ipk.mesh.assign_mesh_level(mesh_level_dict)
    ipk.mesh.meshoperations[0].props.update(OrderedDict({
        "MaxLevel": "3",
        "MinLevel": "2",
        "Local Mesh Parameters Enabled": True,
        "X count": "2",
        "Y count": "2",
        "Z count": "2"
    }))
    ipk.mesh.meshoperations[0].props.pop("Level")
    ipk.mesh.meshoperations[0].update()
```

I would like to know how to modify the code to avoid this error. Looking forward to your reply.

**Reply by pruthvi1991** [reply]
Hi,

I ran into a similar issue when modifying mesh region properties.

The following works for me when trying to assign mesh region properties from a dictionary. Sharing in case this might help you with the mesh operation settings.

```
global_mesh_settings = {'MaxElementSizeZ': '50um', 
                       'MaxElementSizeX': '50um', 
                       'MaxElementSizeY': '50um', 
                       'MinGapX': '1um',
                       'MinGapY': '1um', 
                       'MinGapZ': '1um',
                       'MaxLevels': '1',
                       'NoOGrids': True,
                       'EnforeMLMType': 'None', 
                       }


glob_msh = ipk.mesh.global_mesh_region

glob_msh.UserSpecifiedSettings = True
for key,value in global_mesh_settings.items():
        glob_msh.UserSpecifiedSettings = True
        if hasattr(glob_msh, key):
            setattr(glob_msh,key,value)

glob_msh.update()
```

**Reply by lorenzovecchietti** [reply]
> KeyError: 'Level'

This means that the key has already been removed. 

This is because in your for loop you are always selecting the same mesh operation (so after the first time the "Level" key is not there).

**Reply by aknsj** [reply]
Thank you for your reply. I changed the code to the following and successfully assign mesh operation to multiple solids. 
Thank you very much, your help solved my problem!
```
from collections import OrderedDict
i = 0
for solid_name in ipk.modeler.solid_names:
    if solid_name.lower().startswith("copper"):
        ipk.mesh.assign_mesh_level({solid_name:2})
        ipk.mesh.meshoperations[i].props.update(OrderedDict({
            "MaxLevel": "3",
            "MinLevel": "2",
            "Local Mesh Parameters Enabled": True,
            "X count": "2",
            "Y count": "2",
            "Z count": "2"
        }))
        ipk.mesh.meshoperations[i].props.pop("Level")
        ipk.mesh.meshoperations[i].update()
        i+=1
```

**Reply by lorenzovecchietti** [reply]
Glad I helped :)
